mlir/lib/Dialect/MIOpen/TransformMapBuilder.cpp

```diff
@@ -376,17 +376,43 @@ TransformMapBuilder::operator=(const TransformMapBuilder &other) {
     result = other.result;
     loc = other.loc;
 
-    startIndices = other.startIndices;
     startNames = other.startNames;
     startShape = other.startShape;
-    endIndices = other.endIndices;
     endNames = other.endNames;
     endShape = other.endShape;
     frozen = other.frozen;
+
+    startIndices.clear();
+    for (uint32_t i = 0, e = startNames.size(); i < e; ++i)
+      startIndices.insert({StringRef(startNames[i]), i});
+
+    endIndices.clear();
+    for (const auto &pair : endNames)
+      endIndices.insert({StringRef(pair.second), pair.first});
   }
   return *this;
 }
 
+TransformMapBuilder::TransformMapBuilder(const TransformMapBuilder &other)
+    : b(other.b), result(other.result), loc(other.loc), startIndices(),
+      startNames(other.startNames), startShape(other.startShape), endIndices(),
+      endNames(other.endNames), endShape(other.endShape), frozen(other.frozen) {
+  for (uint32_t i = 0, e = startNames.size(); i < e; ++i)
+    startIndices.insert({StringRef(startNames[i]), i});
+  for (const auto &pair : endNames)
+    endIndices.insert({StringRef(pair.second), pair.first});
+}
+
+TransformMapBuilder::TransformMapBuilder(TransformMapBuilder &&other)
```

Comment:
Why do you have to come up with the additional rvalue referenced constructor? I don't see an instance where you'd have to use move upon them. Plus this looks equivalent with the lvalue version above?

Reply:
I'm adding it in case something calls the move constructor in the future. We don't want to move over dangling pointers (since `TransformMapBuilder` is self-referential)

Reply:
If I remember it right, there's something like:
`TransformMapBuilder::TransformMapBuilder(TransformMapBuilder &&other)=delete;`

Reply:
Which is incorrect. `startIndices` and `endIndices` hold pointers to data within `startNames` and `endNames` - if you naively move/copy the values in `startIndices` or `endIndices`, they'll point to junk data because `startNames` and `endNames` have moved.

Reply:
When it is deleted, you'd not be able to compile if you call move on this class.

Reply:
Right, I saw `=default` and not `=delete`. I'd argue for still *having* a move constructor, since there's nothing here we can't move, but I can delete it instead. I don't have a strong preference on this one.

Reply:
It can go either ways: 
 - use =delete to make sure people cannot move it, or otherwise have to make the implementation right
 - Implement the move constructor correctly, which means you'll have to use move operator on all of the fields, unless the reference case which you customize the implementation to make sure it is right.

I think the existing implementation is deceiving in the sense that: If someone only read the header file and see a move constructor and try to use it to gain performance benefit, the move constructor actually is no different than the lvalue constructor. And actually he/she would pay the price of invoking a `std::move` to do this redundant type conversion.

Reply:
I'd've expected the initialization syntax to do the moves, but apparently not? Will be deleting.